db/migrate/20180116065518_add_hierarchy_paths.rb

```diff
@@ -0,0 +1,45 @@
+#-- encoding: UTF-8
+
+#-- copyright
+# OpenProject is a project management system.
+# Copyright (C) 2012-2017 the OpenProject Foundation (OPF)
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2017 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See doc/COPYRIGHT.rdoc for more details.
+#++
+
+class AddHierarchyPaths < ActiveRecord::Migration[5.0]
+  def change
+    create_table :hierarchy_paths do |t|
+      t.belongs_to :work_package, index: { unique: true }
+      # (255 * 3) = ca 767 bytes is the max length for an index in mysql 5.6 InnoDB
+      t.string :path, null: false, limit: 255
```

Comment:
This will silently fail to insert new paths at production for large primary keys and/or long hierarchy chains. How can we mitigate this?

Reply:
Actually it won't:

```
mysql> INSERT INTO hierarchy_paths (work_package_id, path) VALUES (-4, '111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111');
ERROR 1406 (22001): Data too long for column 'path' at row 1
```

which surprised me as well as I assumed mysql would silently cut off the extra characters and I despised mysql for doing it. Granted, it will fail hard but it will fail.

I thought about implementing a check on changes to the parent but deemed it to be too expansive as I'd have to get the complete branch and check for the length for every possible path.

Given that mysql fails and that 255 characters means that one can have a hierarchy with a depth of 25 while the keys can have 9 digits each at the same time I didn't think it worthwhile. In the long run we can increase the path limit as soon as we restrict mysql to only support newer versions or drop it altogether.